src/platform/silabs/EndpointQueueFilter.cpp

```diff
@@ -0,0 +1,141 @@
+                                          const chip::System::PacketBufferHandle & pktPayload)
+{
+    if (pktInfo.DestPort == kMdnsPort)
+    {
+#if INET_CONFIG_ENABLE_IPV4
+        ip_addr_t mdnsIPv4BroadcastAddr = IPADDR4_INIT_BYTES(224, 0, 0, 251);
+        if (pktInfo.DestAddress == chip::Inet::IPAddress(mdnsIPv4BroadcastAddr))
+        {
+            return FilterOutcome::kAllowPacket;
+        }
+#endif
+        ip_addr_t mdnsIPv6BroadcastAddr = IPADDR6_INIT_HOST(0xFF020000, 0, 0, 0xFB);
+        if (pktInfo.DestAddress == chip::Inet::IPAddress(mdnsIPv6BroadcastAddr))
+        {
+            return FilterOutcome::kAllowPacket;
+        }
+    }
+    return FilterOutcome::kDropPacket;
+}
+
+FilterOutcome HostNameFilter::Filter(const void * endpoint, const IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload)
+{
+    // Drop the mDNS packets which don't contain 'matter' or '<device-hostname>'.
+    const uint8_t matterBytes[] = { 'm', 'a', 't', 't', 'e', 'r' };
+    if (PayloadContains(pktPayload, ByteSpan(matterBytes)) || PayloadContainsCaseInsensitive(pktPayload, ByteSpan(mHostName)))
+    {
+        return FilterOutcome::kAllowPacket;
+    }
+    return FilterOutcome::kDropPacket;
+}
+
+CHIP_ERROR HostNameFilter::SetHostName(const chip::CharSpan & name)
+{
+    ReturnErrorCodeIf(name.size() != sizeof(mHostName), CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnErrorCodeIf(!IsValidMdnsHostName(name), CHIP_ERROR_INVALID_ARGUMENT);
+    memcpy(mHostName, name.data(), name.size());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR HostNameFilter::SetMacAddr(const chip::ByteSpan & mac_addr)
+{
+    ReturnErrorCodeIf(mac_addr.size() < 6, CHIP_ERROR_INVALID_ARGUMENT);
+    const uint8_t * p = mac_addr.data();
+    snprintf((char *) mHostName, sizeof(mHostName), "%02x%02x%02x%02x%02x%02x", p[0], p[1], p[2], p[3], p[4], p[5]);
+    return CHIP_NO_ERROR;
+}
+
+namespace SilabsEndpointQueueFilter {
+
+EndpointQueueFilter::EndpointQueueFilter() : mTooManyFilter(kDefaultAllowedQueuedPackets) {}
+
+EndpointQueueFilter::EndpointQueueFilter(size_t maxAllowedQueuedPackets) : mTooManyFilter(maxAllowedQueuedPackets) {}
+
+FilterOutcome EndpointQueueFilter::FilterBeforeEnqueue(const void * endpoint, const IPPacketInfo & pktInfo,
+                                                       const chip::System::PacketBufferHandle & pktPayload)
+{
+    VerifyOrReturnError(FilterOutcome::kAllowPacket == mTooManyFilter.FilterBeforeEnqueue(endpoint, pktInfo, pktPayload),
+                        FilterOutcome::kDropPacket);
+    if (FilterOutcome::kAllowPacket == mMdnsFilter.Filter(endpoint, pktInfo, pktPayload))
+    {
+        return FilterOutcome::kAllowPacket;
+    }
+    if (FilterOutcome::kAllowPacket == mHostNameFilter.Filter(endpoint, pktInfo, pktPayload))
+    {
+        return FilterOutcome::kAllowPacket;
+    }
+    return FilterOutcome::kAllowPacket;
```

Comment:
This will never drop anything.

The exact order of checks has to be:

```
    if (FilterOutcome::kAllowPacket != mMdnsFilter.Filter(endpoint, pktInfo, pktPayload))
    {
        return FilterOutcome::kDropPacket;
    }
    if (FilterOutcome::kAllowPacket != mHostNameFilter.Filter(endpoint, pktInfo, pktPayload))
    {
        return FilterOutcome::kDropPacket;
    }
    return FilterOutcome::kAllowPacket;
```

Reply:
refactored the logic, thanks!